Mathlib/Logic/Equiv/Fin/Basic.lean

```diff
@@ -247,6 +247,49 @@ theorem finSumFinEquiv_symm_apply_natAdd (x : Fin n) :
 theorem finSumFinEquiv_symm_last : finSumFinEquiv.symm (Fin.last n) = Sum.inr 0 :=
   finSumFinEquiv_symm_apply_natAdd 0
 
+/-- Equivalence between `Fin n ⊕ ℕ` and `ℕ` that sends `inl (a : Fin n)` to
+  `(a : ℕ)` and `inr a` to `a + n`. -/
+def finSumNatEquiv (n : ℕ) : Fin n ⊕ ℕ ≃ ℕ where
+  toFun := Sum.elim Fin.val (· + n)
+  invFun i := if hi : i < n then Sum.inl ⟨i, hi⟩ else Sum.inr (i - n)
+  left_inv i := i.casesOn
+    (fun _ => dif_pos (Fin.is_lt _))
+    (fun _ => (dif_neg (n.le_add_left _).not_gt).trans <| congrArg _ <| Nat.add_sub_cancel _ _)
+  right_inv i := (apply_dite _ _ _ _).trans <| (i.lt_or_ge n).by_cases
+    (fun hi => dif_pos hi)
+    (fun hi => (dif_neg hi.not_gt).trans <| Nat.sub_add_cancel hi)
+
+@[simp] theorem finSumNatEquiv_apply_left (i : Fin n) :
+    (finSumNatEquiv n) (Sum.inl i) = i := rfl
+
+@[simp] theorem finSumNatEquiv_apply_right (i : ℕ) :
+    (finSumNatEquiv n) (Sum.inr i) = i + n := rfl
+
+@[simp] theorem finSumNatEquiv_symm_apply_of_lt {i : ℕ} (hi : i < n) :
+    (finSumNatEquiv n).symm i = Sum.inl ⟨i, hi⟩ := dif_pos hi
+
+@[simp] theorem finSumNatEquiv_symm_apply_of_ge {i : ℕ} (hi : n ≤ i) :
+    (finSumNatEquiv n).symm i = Sum.inr (i - n) := dif_neg (Nat.not_lt_of_ge hi)
+
+theorem finSumNatEquiv_symm_apply_fin (i : Fin n) :
+    (finSumNatEquiv n).symm i = Sum.inl i := by simp
+
+theorem finSumNatEquiv_symm_apply_add_left (i : ℕ) :
+    (finSumNatEquiv n).symm (i + n) = Sum.inr i := by simp
+
+theorem finSumNatEquiv_symm_apply_add_right (i : ℕ) :
+    (finSumNatEquiv n).symm (n + i) = Sum.inr i := by simp
+
+theorem isLeft_finSumNatEquiv_symm_apply (i : ℕ) :
+    ((finSumNatEquiv n).symm i).isLeft ↔ i < n :=
```

Comment:
I think this woudl be better as
```suggestion
    ((finSumNatEquiv n).symm i).isLeft = decide (i < n) :=
```
which could be a `simp` lemma, though my opinion is that `isLeft` and `isRight` are generally pretty unpleasant to use at all.